### DISCUSSION TITLE
Call rest api to create html and serve it with a slugifyed url

### DISCUSSION THREAD

**Discussion by blueray453** [rwf2/Rocket#2287] [Questions]
Can someone please give me an working example of :

when i hit `examples.com/my-12th-post` will create view (HTML+CSS) in the server using `api.examples.com/post/12`  (where `api.examples.com/post/12` is a rest api).

Here, `my-12th-post` is the slug of `post/12`'s headline.

Here, I want two files. One will create the Rest API. The other is for SSR (server side rendering) and slugifying the URL.

_json-api.rs_
```
#[get("/post/12")]
fn func1 -> JSON {
    // Output in JSON when request api.examples.com/post/12
}
```
_server-side-renderer.rs_
```
#[get("/my-12th-post")]
fn func_rend_func1 -> HTML {
    // Output HTML using func1 when request /my-12th-post
}
```

**Comment by blackghost1987** [comment]
I'm not sure I follow the whole idea, but to reuse the code in func1, you could simply call func1 from func_rend_func1, and convert the JSON result into the HTML result you need. You probably want to use a template engine to create the HTML, take a look at:
https://rocket.rs/v0.5-rc/guide/responses/#templates

Although I think the best way to do this is not to call func1 directly, but create a private function instead. This new function would generate your custom type with all the data in it, then you could call it from both functions. The custom type can be serialized to JSON to get the first response type, OR passed as context to the template engine to get the HTML, something like this:

```
fn get_data() -> MyDataType {
  ...
}

#[get("/post/12")]
fn func1() -> Json<MyDataType> {
  let data = get_data();
  Json(data)
}

#[get("/my-12th-post")]
fn func_rend_func1() -> Template {
  let data = get_data();
  Template::render("post", &data)
}
```

Note that this is using fixed, hardcoded route URLs, so it only works with exactly "/post/12" and "my-12th-post"! It's not possible in Rocket to dynamically add routes during runtime AFAIK. If you want to handle multiple ID / slug possibilities, not just "12", you'll need dynamic paths, see:
https://rocket.rs/v0.5-rc/guide/requests/#dynamic-paths

You could use the following dynamic paths:
```
#[get("/post/<id>")]
fn func1(id: u32) -> Json<MyDataType>{
  let data = get_data(id);
  Json(data)
}

#[get("/<slug>")]
fn func_rend_func1(slug: &str) -> Template {
  let id = match slug {
    "my-12th-post" => 12,
     .....
  };
  let data = get_data(id);
  Template::render("post", &data)
}
```

As you can see to achieve this, you'll need some logic to get the ID based on the slug. You can simply use a match like I did in the example, if the slugs are known at compile time. Or if the slugs are always formatted like "my-<id>th-post", then you can use some custom logic to get the integer ID from the string slug.

If the slugs can be anything at all and they are defined at runtime, then you'll definitely need a Database to store the relation between slugs and post IDs, so you could search for the proper ID in there based on the slug.

I hope this helps.
